pandas/core/array_algos/masked_reductions.py

```diff
@@ -0,0 +1,47 @@
+"""
+masked_reductions.py is for reduction algorithms using a mask-based approach
+for missing values.
+"""
+
+import numpy as np
+
+from pandas._libs import missing as libmissing
+from pandas.compat.numpy import _np_version_under1p17
+
+from pandas.core.nanops import check_below_min_count
+
+
+def sum(
+    values: np.ndarray, mask: np.ndarray, skipna: bool = True, min_count: int = 0,
+):
+    """
+    Sum for 1D masked array.
+
+    Parameters
+    ----------
+    values : np.ndarray
+        Numpy array with the values (can be of any dtype that support the
+        operation).
+    mask : np.ndarray
+        Boolean numpy array (True values indicate missing values).
+    skipna : bool, default True
+        Whether to skip NA.
+    min_count : int, default 0
+        The required number of valid values to perform the operation. If fewer than
+        ``min_count`` non-NA values are present the result will be NA.
+    """
+    if not skipna:
+        if mask.any():
+            return libmissing.NA
+        else:
+            if check_below_min_count(values.shape, None, min_count):
+                return libmissing.NA
+            return np.sum(values)
+    else:
+        if check_below_min_count(values.shape, mask, min_count):
+            return libmissing.NA
+
+        if _np_version_under1p17:
+            return np.sum(values[~mask])
+        else:
+            return np.sum(values, where=~mask)
```

Comment:
@jorisvandenbossche This seems like a useful pattern for other reductions like min and max (maybe a special case is needed if we have a `min_count`?), wondering if it might make sense to generalize this to those? Could be more performant than using `nanops`, and could also be done in a way that's extensible to string dtype (where `nanops` can't be used)

Reply:
Yes, that's indeed the goal of this PR (I just wanted to limit it to a single function for the initial PR). For min/max, I think it will be better to write a separate function in module since they don't use `min_count`, but `prod` should be able to reuse this. 
I was planning to do follow-up PRs the and of this week, but feel free to pick it up!